File: python/day_1_python/p_2.py
# Part Two
with open("input.txt") as f:
    lines = []
    for line in f:
        cleaned_line = line.rstrip("\n") 
        lines.append(cleaned_line)

digit_dict = {"one": "1",
        "two": "2",
        "three": "3",
        "four": "4",
        "five": "5",
        "six": "6",
        "seven": "7",
        "eight": "8",
        "nine": "9"
        }

total = 0
i = 0

for line in lines:
    digits = []
    for i, c in enumerate(line):
        if c.isdigit():
            digits.append(c)
        for word in digit_dict:
            if line[i:].startswith(word):
                digits.append(str(digit_dict[word]))
    concat = int(digits[0] + digits[-1])
    total += concat

print(total)
#53348

File: c++/day_1_cc/p_2.cc
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <charconv>

// Let's split the code into functions vs p_1
std::map<std::string, int> createDigitMap() {
    
    std::map<std::string, int> digitMap;
    digitMap["one"] = 1;
    digitMap["two"] = 2;  
    digitMap["three"] = 3;
    digitMap["four"] = 4;
    digitMap["five"] = 5;  
    digitMap["six"] = 6;
    digitMap["seven"] = 7;
    digitMap["eight"] = 8;
    digitMap["nine"] = 9; 

    return digitMap;
}

// Extract the numberical digits from the strings
std::vector<char> extractDigits(const std::string& line) {
  std::vector<char> digits;
  std::map<std::string, int> digitMap = createDigitMap();
  
  for (char c : line) {
    if (std::isdigit(c)) {
      digits.push_back(c);
    }
    for (const auto& [word, value]: digitMap) {
        if (line.find(word, c) != std::string::npos) {
            digits.push_back(value);
            c += word.length() -1;
        }
    }
  }
  return digits;
}

// Concatenate first, last digits eg 7, 8 = 78
int concatenateDigits(const std::vector<char>& digits) {
  if (digits.empty()) {
    return 0; // No digits, return 0
  }

  int firstDigit = digits[0] - '0';
  int lastDigit = digits.back() - '0';
  
  return (firstDigit * 10) + lastDigit;
}

int main() {
    std::ifstream infile("input.txt");
    std::vector<std::string> lines;
    std::string line;
    while (std::getline(infile, line)) {
        lines.push_back(line); 
    }
    infile.close();
    
    int total = 0;
    // lines vector now contains the file content  
    for(auto &l : lines) {
        // Vector to hold the digits, chars for now
        std::vector<char> digits = extractDigits(l);
        int concat = concatenateDigits(digits);
        total += concat;
    }
    std::cout << total;
  
    return 0;
}
